core/src/main/scala/ai/h2o/sparkling/backend/utils/H2OClientUtils.scala

```diff
@@ -15,60 +15,92 @@
+  }
 
-    if (conf.isAutoClusterStartUsed) {
-      val expectedSize = conf.clusterSize.get.toInt
-      val discoveredSize = waitForCloudSize(expectedSize, conf.cloudTimeout)
-      if (discoveredSize < expectedSize) {
-        logError(s"Exiting! External H2O cluster was of size $discoveredSize but expected was $expectedSize!!")
-        hc.stop(stopSparkContext = true)
-        throw new RuntimeException("Cloud size " + discoveredSize + " under " + expectedSize);
+  def startH2OClient(hc: H2OContext, conf: H2OConf, nodes: Array[NodeDesc]): NodeDesc = {
+    if (conf.runsInExternalClusterMode) {
+      setClientIp(conf)
+    } else {
+      if (conf.clientIp.isEmpty) {
+        conf.setClientIp(getHostname(SparkEnv.get))
       }
     }
-
-    RestAPIManager(hc).registerAll()
-    H2O.startServingRestApi()
+    if (!(conf.runsInInternalClusterMode && hc.sparkContext.isLocal)) {
+      val args = getH2OClientArgs(conf).toArray
+      val launcherArgs = toH2OArgs(args, nodes)
+      logDebug(s"Arguments used for launching the H2O client node: ${launcherArgs.mkString(" ")}")
+
+      H2OStarter.start(launcherArgs, false)
+
+      if (conf.runsInInternalClusterMode || conf.isAutoClusterStartUsed) {
+        val expectedSize = nodes.length
+        val discoveredSize = waitForCloudSize(expectedSize, conf.cloudTimeout)
+        if (discoveredSize < expectedSize) {
+          logError(s"Exiting! H2O cluster was of size $discoveredSize but expected was $expectedSize!!")
+          hc.stop(stopSparkContext = true)
+          throw new RuntimeException("Cloud size " + discoveredSize + " under " + expectedSize);
+        }
+      }
+      RestAPIManager(hc).registerAll()
+      H2O.startServingRestApi()
+    }
+    NodeDesc(SparkEnv.get.executorId, H2O.SELF_ADDRESS.getHostAddress, H2O.API_PORT)
   }
 
-  protected[external] def launchShellCommand(cmdToLaunch: Seq[String]): Int = {
-    import scala.sys.process._
-    val processOut = new StringBuffer()
-    val processErr = new StringBuffer()
+  /**
+   * Wait for cloud size from the H2O client point of view
+   */
+  private def waitForCloudSize(expectedSize: Int, timeoutInMilliseconds: Long): Int = {
+    val start = System.currentTimeMillis()
+    while (System.currentTimeMillis() - start < timeoutInMilliseconds) {
+      if (H2O.CLOUD.size() >= expectedSize && Paxos._commonKnowledge) {
+        return H2O.CLOUD.size()
+      }
+      try {
+        Thread.sleep(100)
+      } catch {
+        case _: InterruptedException =>
+      }
+    }
+    H2O.CLOUD.size()
+  }
 
-    val proc = cmdToLaunch.mkString(" ").!(ProcessLogger(
-      { msg =>
-        processOut.append(msg + "\n")
-        println(msg)
-      }, {
-        errMsg =>
-          processErr.append(errMsg + "\n")
-          println(errMsg)
-      }))
+  private def setClientIp(conf: H2OConf): Unit = {
+    val clientIp = identifyClientIp(conf.h2oClusterHost.get)
+    if (clientIp.isDefined && conf.clientNetworkMask.isEmpty) {
```

Comment:
Thanks! Will double check

Reply:
Yup, it's on purpose. The condition in ExernalH2OBackend: 
```
  // to mimic the previous behaviour, set the client ip like this only in manual cluster mode when using multi-cast
    if (conf.clientIp.isEmpty && conf.isManualClusterStartUsed && conf.h2oCluster.isEmpty) {
      conf.setClientIp(getHostname(SparkEnv.get))
    }
```
has no effect because we no longer support manual mode with multicast discovery. So this was just removing old code.

But the condition https://github.com/h2oai/sparkling-water/blob/1fa1752964d60783a6dbbf11b7737d73c07313e6/core/src/main/scala/ai/h2o/sparkling/backend/external/ExternalBackendUtils.scala#L91 remains the same